### PR TITLE
Foxy testing

### PR DESCRIPTION
**Pull request by igorrecioh**
Added tests for pendulum_controller package

### CODE REVIEW

**Comment by igorrecioh** [comment]
@LanderU @carlossvg review this, please. Tell me any issue you may have so as to fix it asap and merge

File: pendulum_controller/test/test_pendulum_controller.cpp [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <memory>
+#include <vector>
+#include "rclcpp/rclcpp.hpp"
+#include "pendulum_controller/pendulum_controller_node.hpp"
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller.cpp:19]: Remove not needed headers: rclcpp, node, etc

File: pendulum_controller/test/test_pendulum_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+#include "pendulum2_msgs/msg/joint_command_stamped.hpp"
+#include "pendulum2_msgs/msg/pendulum_teleop.hpp"
+
+using pendulum::pendulum_controller::PendulumControllerNode;
+using pendulum::pendulum_controller::PendulumController;
+
+using lifecycle_msgs::msg::State;
+using lifecycle_msgs::msg::Transition;
+
+class InitNodesTest : public ::testing::Test
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller.cpp:31]: InitNodesTest is not needed here

File: pendulum_controller/test/test_pendulum_controller.cpp [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  static void TearDownTestCase()
+  {
+    (void)rclcpp::shutdown();
+  }
+};
+
+// Test config
+TEST_F(InitNodesTest, test_config) {
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller.cpp:65]: Just use a zero matrix if there is no need to test the controller functionality, this is valid for all other functions as well

File: pendulum_controller/test/test_pendulum_controller.cpp [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+// Test constructor
+TEST_F(InitNodesTest, test_constructor) {
+  PendulumController::Config config({-10.0000, -51.5393, 356.8637, 154.4146});
+  PendulumController controller(config);
+
+  std::vector<double> state = controller.get_state();
+  std::vector<double> reference = controller.get_reference();
+  PendulumController::Config configObj = controller.get_config();
+
+  ASSERT_EQ(state[0], 0.0);
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller.cpp:79]: Create `reference_default` and `state_default` vectors to compare. e.g.

```cpp
std::vector<double> state_default{0.0, 0.0, M_PI, 0.0};
ASSERT_EQ(state[0], state_default[0]);
...
```

File: pendulum_controller/test/test_pendulum_controller.cpp [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+  ASSERT_EQ(state[1], 0.0);
+  ASSERT_EQ(state[2], M_PI);
+  ASSERT_EQ(state[3], 0.0);
+
+  ASSERT_EQ(reference[0], 0.0);
+  ASSERT_EQ(reference[1], 0.0);
+  ASSERT_EQ(reference[2], M_PI);
+  ASSERT_EQ(reference[3], 0.0);
+
+  ASSERT_EQ(configObj.get_feedback_matrix(), config.get_feedback_matrix());
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller.cpp:89]: This shouldn't be tested here

File: pendulum_controller/test/test_pendulum_controller.cpp [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+TEST_F(InitNodesTest, test_constructor) {
+  PendulumController::Config config({-10.0000, -51.5393, 356.8637, 154.4146});
+  PendulumController controller(config);
+
+  std::vector<double> state = controller.get_state();
+  std::vector<double> reference = controller.get_reference();
+  PendulumController::Config configObj = controller.get_config();
+
+  ASSERT_EQ(state[0], 0.0);
+  ASSERT_EQ(state[1], 0.0);
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller.cpp:80]: Use Floating-Point Macros

File: pendulum_controller/test/test_pendulum_controller.cpp [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+  ASSERT_EQ(state[3], 2.0);
+}
+
+// Test update command
+TEST_F(InitNodesTest, test_update_command) {
+  PendulumController::Config config({-10.0000, -51.5393, 356.8637, 154.4146});
+  PendulumController controller(config);
+
+  pendulum2_msgs::msg::JointCommandStamped msg_command;
+  controller.reset();
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller.cpp:136]: ~~Create a test function for the reset~~ Move test_reset before this function

File: pendulum_controller/test/test_pendulum_controller.cpp [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+  ASSERT_FLOAT_EQ(msg_command.cmd.force, 251.731);
+}
+
+// Test reset
+TEST_F(InitNodesTest, test_reset) {
+  PendulumController::Config config({-10.0000, -51.5393, 356.8637, 154.4146});
+  PendulumController controller(config);
+  controller.reset();
+
+  std::vector<double> ref = controller.get_reference();
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller.cpp:162]: Create reference_default and state_default vectors

File: pendulum_controller/test/test_pendulum_controller.cpp [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+  msg_status.velocity.emplace_back(2.0);
+  controller.update_status_data(msg_status);
+
+  pendulum2_msgs::msg::PendulumTeleop msg_teleop;
+  msg_teleop.cart_position = M_PI / 2;
+  msg_teleop.cart_velocity = 3.0;
+  controller.update_teleop_data(msg_teleop);
+
+  controller.update_command_data(msg_command);
+  ASSERT_FLOAT_EQ(msg_command.cmd.force, 251.731);
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller.cpp:153]: I would create arrays `msg_status`, `msg_command`, and `msg_teleop` with values taken from simulation to check several iterations. For example, we could use [this](https://gitlab.com/carlossvg/inverted_pendulum_sim) to collect some simulated values.

File: pendulum_controller/test/test_pendulum_controller.cpp [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+  std::vector<double> state = controller.get_state();
+  ASSERT_EQ(state[0], 0.0);
+  ASSERT_EQ(state[1], 0.0);
+  ASSERT_EQ(state[2], M_PI);
+  ASSERT_EQ(state[3], 0.0);
+}
+
+// Test calculate
+TEST(CalculateTest, test_calculate) {
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller.cpp:176]: remove this

File: pendulum_controller/test/test_pendulum_controller_node.cpp [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+  static void TearDownTestCase()
+  {
+    (void)rclcpp::shutdown();
+  }
+};
+
+TEST_F(InitNodesTest, test_options_constructor) {
+  node_options.parameter_overrides(params);
+
+  const auto test_node =
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller_node.cpp:62]: Just test the constructor does not throw with `ASSERT_NO_THROW` and test any possible bad configuration if there is any

File: pendulum_controller/test/test_pendulum_controller_node.cpp [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+      rclcpp_lifecycle::Transition(Transition::TRANSITION_DEACTIVATE)).id());
+  ASSERT_EQ(
+    State::PRIMARY_STATE_UNCONFIGURED, test_node->trigger_transition(
+      rclcpp_lifecycle::Transition(Transition::TRANSITION_CLEANUP)).id());
+  ASSERT_EQ(
+    State::PRIMARY_STATE_FINALIZED, test_node->trigger_transition(
+      rclcpp_lifecycle::Transition(Transition::TRANSITION_UNCONFIGURED_SHUTDOWN)).id());
+}
+
+TEST_F(InitNodesTest, test_param_constructor) {
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller_node.cpp:99]: Move this test before test_transition

File: pendulum_controller/test/test_pendulum_controller_node.cpp [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  node_options.parameter_overrides(params);
+
+  const auto test_node_ptr =
+    std::make_shared<PendulumControllerNode>(node_options);
+  auto names = test_node_ptr->get_node_names();
+
+  // EXPECT_EQ(names.size(), 1u);
+  // EXPECT_STREQ(names[0].c_str(), "/pendulum_controller");
+  // EXPECT_STREQ("/", test_node_ptr->get_namespace());
+}
```
Comment (carlossvg) [inline pendulum_controller/test/test_pendulum_controller_node.cpp:109]: Add a test to check all the required params exist. e.g. https://gitlab.com/autowarefoundation/autoware.auto/AutowareAuto/-/blob/0d12a28fe7099a84bacf953c6b6b6093b372aba9/src/perception/segmentation/euclidean_cluster_nodes/test/test_euclidean_cluster_nodes.hpp#L41-106

**Pull request closed without merging**